src/FreeAge/common/type_info.cpp

```diff
@@ -0,0 +1,621 @@
+  );
+
+  // Villagers have +5 attack against Wild Boars and carry +15 food from hunting.
+  age[0].addModification(
+    FilterOfUnitsByArmor(DamageType::Villager), // easier that filter by type
+    Modification(Type::Damage, Operation::Add, 5, DamageType::Boar)
+  );
+  age[0].addModification(
+    ObjectTypeFilter::Civilization,
+    Modification(Type::VillagerCarryingCapacity, Operation::Add, 15, UnitType::FemaleHunter)
+  );  // repeat for UnitType::MaleHunter
+
+  // +10 population cap in the Imperial Age.
+  age[3].addModification(
+    ObjectTypeFilter::Civilization,
+    Modification(Type::PopulationCap, Operation::Add, 10)
+  );
+
+  // Team bonus: Barracks work 20% faster.
+  age[3].teamBonus.addModification(
+    FilterOfBuildingsByType(BuildingType::Barracks),
+    Modification(Type::WorkRate, Operation::MultAdd, 20) // or use Mult with 1.2f
+  )
+
+}
+
+// Unable to hanlde cases (yet):
+// - Ethiopians: Receive +100 food and +100 gold whenever a new Age is reached.
+// - Lithuanians: Each garrisoned Relic gives +1 attack to Knights and Leitis (maximum +4).
+// - Tatars: Herdables contain +50% food. (how this is even work?)
+//      (could be a boolean value at civ category)
+// - Celts: Enemy herdables can be converted regardless of enemy units next to them.
+// - Burmese: Relics are visible on the map from the game start.
+// - Khmer: Khmer Villagers drop off food from farms (1 unit at a time) without needing a drop site
+// - Vietnamese: Reveal enemy positions at game start.
+//
+// TODO: fill the list
+
+
+// And how about Feitoria?
+
+void Simulation() {
+  // ...
+  if (onOneSecondTick && player.stats.GetActiveBuildingTypeCount(BuildingType::Feitoria) > 0) { // a single array access
+    player.resources.add(player.stats.GetActiveBuildingTypeCount(BuildingType::Feitoria) * FeitoriaGeneration);
+  }
+  // ...
+}
+
+
+// ===================================================================================
+// === Unit Conversion
+// ===================================================================================
+
+// Here is a solution for the unit conversion problem under this design
+
+struct Unit {
+
+  Player* player;
+  UnitType type;
+
+  UnitTypeInfo *info;
+
+  std::unordered_map<UnitType, UnitTypeInfo>* unitTypes = nullptr;
+  // TODO: keep a copy of the CivilizationStats ? is it worth it ?
+
+  // for not converted units the cost of this implemention
+  // time: one pointer comperation per UnitType change (~0 per frame)
+
+  Unit(UnitType type, ....., Player* sourcePlayer = nullptr) {
+
+    if (sourcePlayer == nullptr) {
+      info = player->GetUnitTypeInfo(type); // pointer valid until end of time
+    } else {
+      // this is a converted unit
+
+      // GenerateStaticUnitTypeInfosNeededBy create a new map with copies of the
+      // all the UnitTypeInfo that are needed by the given type
+      allunittypes = sourcePlayer->GenerateStaticUnitTypeInfosNeededBy(type);
```

Comment:
Even with a sub-type system you would still need the UnitTypeStats for all the possible sub-types at the time of the conversion.

Reply:
The idea would be that only a single UnitTypeStats exists for each primary type and thus all sub-types use the same UnitTypeStats (since for example Gold and Stone Miners should not differ in their stats). However, that only works as long as there is no need to use specific stats for specific sub-types.

Reply:
The stats differ in most cases eg:
- Trebuchet packed and upacked has different armor (and attack obviously)
- Monk with and without a relic have different radius (used to also have different armor)
- The hunter villager has a range attack instead of melee

The field workRate for each type o Villager models the gathering rate, but this could easily be put in the CivilizationStats.

Maybe there are more?